File: backend/app/schemas/recovery.py
from typing import Optional, List, Dict, Any
from enum import Enum
from datetime import datetime
from pydantic import BaseModel, Field


class RootCauseCategory(str, Enum):
    NETWORK_TIMEOUT = "NETWORK_TIMEOUT"
    GATEWAY_ERROR = "GATEWAY_ERROR"
    INSUFFICIENT_FUNDS = "INSUFFICIENT_FUNDS"
    USER_DROPPED = "USER_DROPPED"
    CARD_LIMIT_EXCEEDED = "CARD_LIMIT_EXCEEDED"
    FRAUD_DECLINE = "FRAUD_DECLINE"
    SYSTEM_DOWN = "SYSTEM_DOWN"
    AUTHENTICATION_FAILURE = "AUTHENTICATION_FAILURE"
    EXPIRED_CARD = "EXPIRED_CARD"
    OTHER = "OTHER"


class ActionType(str, Enum):
    RETRY_NOW = "RETRY_NOW"
    RETRY_LATER = "RETRY_LATER"
    PAYMENT_UPDATE = "PAYMENT_UPDATE"
    REMINDER = "REMINDER"
    ESCALATE = "ESCALATE"
    STOP = "STOP"


class ExecutionMode(str, Enum):
    RAZORPAY_TEST = "RAZORPAY_TEST"
    SIMULATION = "SIMULATION"
    DRY_RUN = "DRY_RUN"


class CaseStatus(str, Enum):
    DETECTED = "DETECTED"
    DIAGNOSED = "DIAGNOSED"
    DECISION_READY = "DECISION_READY"
    APPROVED = "APPROVED"
    EXECUTING = "EXECUTING"
    RECOVERED = "RECOVERED"
    WAITING = "WAITING"
    STOPPED = "STOPPED"
    ESCALATED = "ESCALATED"
    FAILED = "FAILED"


class OutcomeType(str, Enum):
    RECOVERED = "RECOVERED"
    FAILED_RETRYABLE = "FAILED_RETRYABLE"
    FAILED_TERMINAL = "FAILED_TERMINAL"
    NO_RESPONSE = "NO_RESPONSE"
    DISMISSED = "DISMISSED"
    OPTED_OUT = "OPTED_OUT"
    ESCALATED_MANUALLY = "ESCALATED_MANUALLY"


class CasePriority(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class AIDiagnosisSource(str, Enum):
    LLM = "LLM"
    DETERMINISTIC_FALLBACK = "DETERMINISTIC_FALLBACK"


class AIDiagnosisResult(BaseModel):
    diagnosis: str
    failure_type: RootCauseCategory
    recommended_action: ActionType
    reason: str
    confidence: float = Field(..., ge=0.0, le=1.0)
    source: AIDiagnosisSource


# -----------------------------------------------------------------------------
# Pipeline Step Schemas
# -----------------------------------------------------------------------------

class DiagnosisResult(BaseModel):
    root_cause_category: RootCauseCategory
    confidence_score: float = Field(..., ge=0.0, le=1.0)
    reasoning: str
    recommended_actions: List[ActionType]
    is_terminal_decline: bool = False
    diagnostic_details: Dict[str, Any] = Field(default_factory=dict)
    ai_diagnosis: Optional[AIDiagnosisResult] = None


class ActionCandidate(BaseModel):
    action_type: ActionType
    score: float
    confidence: float
    historical_success_rate: float
    historical_attempts: int
    reasoning: str


class PolicyDecision(BaseModel):
    selected_action: ActionType
    confidence_score: float
    reasoning: str
    ranked_candidates: List[ActionCandidate]
    used_historical_learning: bool = False


class GuardrailResult(BaseModel):
    allowed: bool
    proposed_action: ActionType
    final_action: ActionType
    is_overridden: bool = False
    override_reason: Optional[str] = None
    checks_passed: Dict[str, bool] = Field(default_factory=dict)
    details: Dict[str, Any] = Field(default_factory=dict)


class ExecutionResult(BaseModel):
    action_id: str
    action_type: ActionType
    execution_mode: ExecutionMode
    status: str
    payload: Dict[str, Any] = Field(default_factory=dict)
    guardrail_check_passed: bool = True
    executed_at: datetime = Field(default_factory=datetime.utcnow)


class OutcomeResult(BaseModel):
    outcome_id: str
    outcome_type: OutcomeType
    is_successful: bool
    recovered_amount: float
    new_payment_id: Optional[str] = None
    recovery_time_seconds: Optional[int] = None
    response_payload: Dict[str, Any] = Field(default_factory=dict)


class LearningUpdateResult(BaseModel):
    merchant_id: str
    root_cause_category: str
    action_type: str
    total_attempts: int
    successful_recoveries: int
    success_rate: float
    total_recovered_amount: float


# -----------------------------------------------------------------------------
# API Request & Response Schemas
# -----------------------------------------------------------------------------

class RecoveryActionResponse(BaseModel):
    id: str
    case_id: str
    action_type: ActionType
    execution_mode: ExecutionMode
    status: str
    sequence_number: int
    scheduled_at: Optional[datetime] = None
    executed_at: Optional[datetime] = None
    payload: Dict[str, Any] = Field(default_factory=dict)
    guardrail_check_passed: bool
    ai_confidence_score: Optional[float] = None
    ai_reasoning: Optional[str] = None
    created_at: datetime


class RecoveryOutcomeResponse(BaseModel):
    id: str
    case_id: str
    action_id: str
    outcome_type: OutcomeType
    is_successful: bool
    recovered_amount: float
    new_payment_id: Optional[str] = None
    recovery_time_seconds: Optional[int] = None
    recorded_at: datetime


class AuditLogResponse(BaseModel):
    id: str
    case_id: Optional[str] = None
    action_id: Optional[str] = None
    actor_type: str
    actor_id: Optional[str] = None
    event_type: str
    severity: str
    description: str
    details: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class RecoveryCaseResponse(BaseModel):
    id: str
    merchant_id: str
    payment_id: str
    customer_id: Optional[str] = None
    customer_name: Optional[str] = None
    customer_email: Optional[str] = None
    amount: float
    currency: str = "INR"
    status: CaseStatus
    priority: CasePriority
    retry_count: int
    communication_count: int
    recovered_amount: float
    next_action_due_at: Optional[datetime] = None
    diagnosis_summary: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime
    resolved_at: Optional[datetime] = None


class CaseDetailResponse(RecoveryCaseResponse):
    actions: List[RecoveryActionResponse] = Field(default_factory=list)
    outcomes: List[RecoveryOutcomeResponse] = Field(default_factory=list)
    audit_logs: List[AuditLogResponse] = Field(default_factory=list)


class BatchCreateRequest(BaseModel):
    name: str = "Demo Recovery Batch"
    description: Optional[str] = "Batch of failed payments for adaptive recovery"
    seed_synthetic_count: int = Field(default=6, ge=1, le=50)


class BatchRunResponse(BaseModel):
    batch_id: str
    merchant_id: str
    name: str
    status: str
    total_records: int
    processed_records: int
    recovered_records: int
    total_amount_at_risk: float
    total_recovered_amount: float
    recovery_rate: float
    baseline_recovered_amount: float
    incremental_revenue: float
    recovery_lift_percentage: float
    cases: List[RecoveryCaseResponse] = Field(default_factory=list)


class RecoveryMetricsResponse(BaseModel):
    merchant_id: str
    total_revenue_at_risk: float
    recoveriq_recovered_revenue: float
    baseline_recovered_revenue: float
    incremental_revenue_recovered: float
    recovery_lift_percentage: float
    total_cases_processed: int
    total_cases_recovered: int
    overall_recovery_rate: float
    success_rate_by_category: Dict[str, float] = Field(default_factory=dict)
    top_recovery_actions: List[Dict[str, Any]] = Field(default_factory=list)

File: frontend/src/types/recovery.ts
export type CaseStatus =
  | 'DETECTED'
  | 'DIAGNOSED'
  | 'DECISION_READY'
  | 'APPROVED'
  | 'EXECUTING'
  | 'RECOVERED'
  | 'WAITING'
  | 'STOPPED'
  | 'ESCALATED'
  | 'FAILED';

export type CasePriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export type RootCauseCategory =
  | 'NETWORK_TIMEOUT'
  | 'GATEWAY_ERROR'
  | 'INSUFFICIENT_FUNDS'
  | 'USER_DROPPED'
  | 'CARD_LIMIT_EXCEEDED'
  | 'FRAUD_DECLINE'
  | 'SYSTEM_DOWN'
  | 'AUTHENTICATION_FAILURE'
  | 'EXPIRED_CARD'
  | 'OTHER';

export type ActionType =
  | 'RETRY_NOW'
  | 'RETRY_LATER'
  | 'PAYMENT_UPDATE'
  | 'REMINDER'
  | 'ESCALATE'
  | 'STOP';

export type OutcomeType =
  | 'RECOVERED'
  | 'FAILED_RETRYABLE'
  | 'FAILED_TERMINAL'
  | 'NO_RESPONSE'
  | 'DISMISSED'
  | 'OPTED_OUT'
  | 'ESCALATED_MANUALLY';

export interface RecoveryMetrics {
  merchant_id: string;
  total_revenue_at_risk: number;
  recoveriq_recovered_revenue: number;
  baseline_recovered_revenue: number;
  incremental_revenue_recovered: number;
  recovery_lift_percentage: number;
  total_cases_processed: number;
  total_cases_recovered: number;
  overall_recovery_rate: number;
  success_rate_by_category: Record<string, number>;
  top_recovery_actions: Array<{
    action: string;
    category: string;
    success_rate: number;
    attempts: number;
    total_recovered: number;
  }>;
}

export interface AIDiagnosisData {
  diagnosis?: string;
  failure_type?: RootCauseCategory | string;
  recommended_action?: ActionType | string;
  reason?: string;
  confidence?: number;
  source?: 'LLM' | 'DETERMINISTIC_FALLBACK' | string;
}

export interface DiagnosisSummary {
  root_cause_category?: RootCauseCategory | string;
  confidence_score?: number;
  reasoning?: string;
  recommended_actions?: string[];
  is_terminal_decline?: boolean;
  diagnosed_at?: string;
  details?: Record<string, unknown>;
  source?: 'LLM' | 'DETERMINISTIC_FALLBACK' | string;
  failure_type?: RootCauseCategory | string;
  recommended_action?: ActionType | string;
  diagnosis?: string;
  reason?: string;
  confidence?: number;
  ai_diagnosis?: AIDiagnosisData;
}

export interface RecoveryCase {
  id: string;
  merchant_id: string;
  payment_id: string;
  customer_id?: string | null;
  customer_name?: string | null;
  customer_email?: string | null;
  amount: number;
  currency: string;
  status: CaseStatus;
  priority: CasePriority;
  retry_count: number;
  communication_count: number;
  recovered_amount: number;
  diagnosis_summary: DiagnosisSummary;
  created_at: string;
  updated_at: string;
  resolved_at?: string | null;
}

export interface RecoveryAction {
  id: string;
  case_id: string;
  action_type: ActionType | string;
  execution_mode: string;
  status: string;
  sequence_number: number;
  scheduled_at?: string | null;
  executed_at?: string | null;
  payload: Record<string, unknown>;
  guardrail_check_passed: boolean;
  ai_confidence_score?: number | null;
  ai_reasoning?: string | null;
  created_at: string;
}

export interface RecoveryOutcome {
  id: string;
  case_id: string;
  action_id: string;
  outcome_type: OutcomeType | string;
  is_successful: boolean;
  recovered_amount: number;
  new_payment_id?: string | null;
  recovery_time_seconds?: number | null;
  recorded_at: string;
}

export interface AuditLog {
  id: string;
  case_id?: string | null;
  action_id?: string | null;
  actor_type: string;
  actor_id?: string | null;
  event_type: string;
  severity: string;
  description: string;
  details: Record<string, unknown>;
  created_at: string;
}

export interface CaseDetail extends RecoveryCase {
  actions: RecoveryAction[];
  outcomes: RecoveryOutcome[];
  audit_logs: AuditLog[];
}

export interface BatchRunResult {
  batch_id: string;
  merchant_id: string;
  name: string;
  status: string;
  total_records: number;
  processed_records: number;
  recovered_records: number;
  total_amount_at_risk: number;
  total_recovered_amount: number;
  recovery_rate: number;
  baseline_recovered_amount: number;
  incremental_revenue: number;
  recovery_lift_percentage: number;
  cases: RecoveryCase[];
}
